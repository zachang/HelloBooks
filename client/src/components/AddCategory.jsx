import React from 'react';
import classnames from 'classnames';
import PropTypes from 'react-proptypes';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import AdminHeader from './common/AdminHeader.jsx';
import AdminSidebar from './common/AdminSidebar.jsx';
import { addCategoryAction } from '../actions/categoryAction';


export class AddCategory extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      categoryData: {
        category_name: '',
      },
      errors: null,
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ errors: nextProps.categoryState.errors });
  }

  handleChange(e) {
    const categoryData = this.state.categoryData;
    categoryData[e.target.name] = e.target.value;
    this.setState({ categoryData })
  }

  handleSubmit(e) {
    e.preventDefault();
    this.props.addCategoryAction(this.state.categoryData);
  }

  render() {
    const { categoryData } = this.state;

    return (
      <div className="row">
        <AdminHeader/>
        <AdminSidebar/>
        <div className="container mainCon" style={{ marginLeft: '5%' }}>
          <div className="row">
            <div className="col s10 m8 l6 bookcat" style={{ marginLeft: '35%', marginTop: '20%' }}>

              <div className="row">
                <form className="col s10" onSubmit={this.handleSubmit}>
                  <div className="row">
                    <div className="input-field col s12">
                      <input
                        id='category_name'
                        name='category_name'
                        type='text'
                        className={ classnames({
                          'invalid': (this.state.errors && !!this.state.errors['category_name'])? !!this.state.errors['category_name'] : false
                        }) }
                        value={categoryData.category_name}
                        onChange={ this.handleChange }
                      />
                      <label
                        htmlFor='category'
                        className={(this.state.errors && !!this.state.errors['category_name'] || categoryData.category_name.length > 0) ?
                          'custom-active custom-validate' : 'custom-validate'}
                        data-error={(this.state.errors && !!this.state.errors['category_name']) ? this.state.errors['category_name'] : ''}
                      >
                        Category
                      </label>
                    </div>
                  </div>

                  <div className="row">
                    <div className="col m12 s12">
                      <button
                        className='col s12 btn btn-large waves-effect waves-light'
                        type='submit'
                        name='action'> Add Category
                      </button>

                      <div style={{ color: 'red', float: 'right' }}>{this.props.categoryState.fails}</div>
                    </div>
                  </div>
                </form>
              </div>

            </div>
          </div>

        </div>
      </div>

    );
  }
}

AddCategory.propTypes = {
  categoryState: PropTypes.object.isRequired,
  addCategoryAction: PropTypes.func.isRequired
};
const mapStateToProps = state => ({
  categoryState: state.categoryReducer,
  tokenState: state.tokenReducer
});
const mapDispatchToProps = dispatch =>
  bindActionCreators({ addCategoryAction }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(AddCategory);