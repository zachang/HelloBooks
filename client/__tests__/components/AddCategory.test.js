import React from 'react';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import storeConfiguration from '../../src/store/store';
import { AddCategory } from '../../src/components/AddCategory.jsx';

const event = {
  preventDefault: jest.fn(),
  target: {
    value: 'mmmmmm'
  }
};

const categoryData = {
  categoryName: '',
};

describe('<AddCategory/>', () => {
  // create a spy function for addCategoryAction
  const addCategoryAction = sinon.spy();
  // spy on handleSubmit of AddCategory Page
  sinon.spy(AddCategory.prototype, 'handleSubmit');
  // spy on handleChange of AddCategory Page
  sinon.spy(AddCategory.prototype, 'handleChange');
  // spy on componentWillReceiveProps of AddCategory Page
  sinon.spy(AddCategory.prototype, 'componentWillReceiveProps');

  const showToast = true;
  const props = {
    categoryState: {
      success: null,
      errors: null,
      fails: null
    },
    showToast,
    addCategoryAction
  };

  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  let shallowComponent;
  beforeEach(() => {
    store = mockStore(storeConfiguration);
    shallowComponent = shallow(<AddCategory {...props} />);
  });

  it('renders <AddCategory /> component', () => {
    const tree = toJson(shallowComponent);
    expect(tree).toMatchSnapshot();
  });

  it('Should check that addCategoryAction is called', () => {
    shallowComponent.find('form').simulate('submit', { preventDefault() {} });
    expect(addCategoryAction.called).toBe(true);
  });

  it('should check that handleChange is called', () => {
    shallowComponent.instance().handleChange(event);
    expect(AddCategory.prototype.handleChange.called).toEqual(true);
  });

  it('calls componentWillReceiveProps category created', () => {
    props.categoryState.success = true;
    shallowComponent.setState({
      showToast
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(AddCategory.prototype.componentWillReceiveProps.called)
      .toEqual(true);
  });

  it('calls componentWillReceiveProps book not created', () => {
    props.categoryState.errors = 'Category not added!';
    props.categoryState.success = false;
    shallowComponent.setState({
      errors: props.categoryState.errors,
      showToast
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(AddCategory.prototype.componentWillReceiveProps.called)
      .toEqual(true);
  });
});