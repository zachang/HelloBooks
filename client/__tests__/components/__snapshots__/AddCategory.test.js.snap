// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCategory/> renders <AddCategory /> component 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col s10 m8 l6 bookcat"
      style={
        Object {
          "marginLeft": "35%",
          "marginTop": "20%",
        }
      }
    >
      <div
        className="row"
      >
        <form
          className="col s10"
          onSubmit={[Function]}
        >
          <div
            className="row"
          >
            <div
              className="input-field col s12"
            >
              <input
                className=""
                id="categoryName"
                name="categoryName"
                onChange={[Function]}
                type="text"
                value=""
              />
              <label
                className="custom-validate"
                data-error=""
                htmlFor="category"
              >
                Category
              </label>
            </div>
          </div>
          <div
            className="row"
          >
            <div
              className="col m12 s12"
            >
              <button
                className="col s12 btn btn-large waves-effect waves-light"
                name="action"
                type="submit"
              >
                 Add Category
              </button>
              <div
                style={
                  Object {
                    "color": "red",
                    "float": "right",
                  }
                }
              />
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
